### ISSUE TITLE
ini_manage state shows changes when there are none

### ISSUE THREAD

**Issue by syphernl** [saltstack/salt#21219]
The `ini_manage` state doesn't appear to be checking the current state and is always returning changes even when there are none and its currently in the right state.

Example:
```
----------
          ID: /etc/mysql/conf.d/customized.cnf
    Function: ini.options_present
      Result: True
     Comment: Changes take effect
     Started: 14:15:02.345800
    Duration: 27.81 ms
     Changes:   
              ----------
              changes:
                  ----------
                  mysqld:
                      ----------
                      long_query_time:
                          ----------
                          after:
                              3
                          before:
                              3
                      query_cache_type:
                          ----------
                          after:
                              1
                          before:
                              1
                      slow_query_log:
                          ----------
                          after:
                              1
                          before:
                              1
                      table_cache:
                          ----------
                          after:
                              1024
                          before:
                              1024
                      table_open_cache:
                          ----------
                          after:
                              524288
                          before:
                              524288
              file_name:
                  /etc/mysql/conf.d/customized.cnf
```

Similarly to the `file` state it should return something along the lines of: 
```
 	File /etc/mysql/conf.d/customized.cnf is in the correct state
```

Minion versions:
```
           Salt: 2014.7.0
         Python: 2.7.6 (default, Mar 22 2014, 22:59:56)
         Jinja2: 2.7.2
       M2Crypto: 0.21.1
 msgpack-python: 0.3.0
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: Not Installed
         PyYAML: 3.10
          ioflo: Not Installed
          PyZMQ: 14.0.1
           RAET: Not Installed
            ZMQ: 4.0.4
           Mako: 0.9.1
```

**Comment by syphernl** [comment]
Pinging @Akilesh1597 as he is the maintainer/initial committer :)

**Comment by jfindlay** [comment]
@syphernl, thanks for the report.  Does this also happen on the latest supported release, `2014.7.2`?

**Comment by jfindlay** [comment]
These are the changes in that file between the versions:
```diff
$ git diff v2014.7.0 v2014.7.2 -- salt/states/ini_manage.py
diff --git a/salt/states/ini_manage.py b/salt/states/ini_manage.py
index b1cff27..cf86fa8 100644
--- a/salt/states/ini_manage.py
+++ b/salt/states/ini_manage.py
@@ -58,8 +58,10 @@ def options_present(name, sections=None):
             current_value = __salt__['ini.get_option'](name,
                                                        section,
                                                        key)
-            if current_value == sections[section][key]:
+            # Test if the change is necessary
+            if current_value == str(sections[section][key]):
                 continue
+
             ret['changes'] = __salt__['ini.set_option'](name,
                                                         sections)
             if 'error' in ret['changes']:
@@ -206,8 +208,8 @@ class _DictDiffer(object):
     def __init__(self, current_dict, past_dict):
         self.current_dict = current_dict
         self.past_dict = past_dict
-        self.set_current = set(current_dict.keys())
-        self.set_past = set(past_dict.keys())
+        self.set_current = set(current_dict)
+        self.set_past = set(past_dict)
         self.intersect = self.set_current.intersection(self.set_past)

     def added(self):
```

@syphernl, this change might fix the problem you're having, or the `options_present` function might need to be updated to use a similar mechanism.

**Comment by syphernl** [comment]
I haven't been able to test that out since there are no Ubuntu packages for `2014.7.2` yet. Perhaps I can add the module and state to the file server for the time being.

**Comment by jfindlay** [comment]
@syphernl, thanks.  Let me know if you have any other updates.

**Comment by jmlowe** [comment]
This module is more broken than just showing changes when there aren't any, it will also show no changes when there should be some and no changes when the file is missing.

**Comment by Akilesh1597** [comment]
Hi guys,
I will look into this and other issues reported on this module. you can expect a better module in two days.

**Comment by stardust85** [comment]
@Akilesh1597 any progress on this?

**Comment by Akilesh1597** [comment]
Please check pull request 26359

**Comment by basepi** [comment]
Fixed in #26359
